### PR TITLE
Implement Azure EventHubs as a TX Log

### LINKED ISSUE #2542: Implement Azure EventHubs as a TX Log
Want to handle a few things under this ticket:
- Implementing Azure Eventhubs as a TX Log. 
  - Implementing the Log interface and all of the related interfaces used by it.
  - Handling auth in the same way as blobstore (defaultazurecredentials)
  - Implementing any integrant related config to be able to use it as a component.
  - Adding any relevant tests against the Log.

### PR DESCRIPTION
**Pull request by danmason**
Within this PR, I make some changes to the structure azure module, and introduce an eventhubs based txlog:
- Split out all of the azure blob object store code, and eventhubs log code, to separate files, keeping the main file for integrant config.
- Added a new dependency to the azure module for the eventhubs client library.
- Integrant "derive key" setup for the new tx log implementation (allows it to replace the `:xtdb/log`)
- Setup an eventhub namespace/eventhub for CI/testing/dev.
- Added it to the azure_testing dev file to play around with.
- Implementing the interface, using `xtdb.log/handle-polling-subscription` to handle the susbcribe behaviour - also implements closable to close consumer/producer.
- Testing the new log implementation with a very basic submit test.
- Adding options to allow the module to manage/create it's own eventhub.
- Documentation on how to use/configure the event hubs log.

## Potential follow up work

The following would likely be the preferable way to handle the subscription, and could likely be done as a piece of followup work:
- Looking into handling subscription using a `notifying-subscriber-handler` and the [azure event processor class? ](https://learn.microsoft.com/en-us/java/api/com.azure.messaging.eventhubs.eventprocessorclientbuilder?view=azure-java-stable#com-azure-messaging-eventhubs-eventprocessorclientbuilder-processevent(java-util-function-consumer(com-azure-messaging-eventhubs-models-eventcontext)))

We could also likely do with some generic TxLog based tests (similar to the Object Store tests) that we can run against our TxLog implementations

### CODE REVIEW

**Comment by danmason** [comment]
Currently seems to get stuck on tests - not entirely sure why. The following error is thrown in the polling-subscription thread:
```
16:10:46.313 [xtdb-tx-subscription-pool-1-thread-2] WARN  xtdb.log - Error polling for txs, will retry
reactor.core.Exceptions$ReactiveException: java.lang.InterruptedException
        at reactor.core.Exceptions.propagate(Exceptions.java:396)
        at reactor.core.publisher.BlockingIterable$SubscriberIterator.hasNext(BlockingIterable.java:184)
        at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:48)
        at clojure.core.protocols$fn__8246.invokeStatic(protocols.clj:75)
        at clojure.core.protocols$fn__8246.invoke(protocols.clj:75)
        at clojure.core.protocols$fn__8194$G__8189__8207.invoke(protocols.clj:13)
        at clojure.core$reduce.invokeStatic(core.clj:6895)
        at clojure.core$mapv.invokeStatic(core.clj:6979)
        at clojure.core$mapv.invoke(core.clj:6979)
        at xtdb.azure.log.EventHubLog.readRecords(log.clj:56)
        at xtdb.log$handle_polling_subscription$fn__33997$fn__34000.invoke(log.clj:52)
        at xtdb.log$handle_polling_subscription$fn__33997.invoke(log.clj:51)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.InterruptedException: null
        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:1638)
        at reactor.core.publisher.BlockingIterable$SubscriberIterator.hasNext(BlockingIterable.java:179)
        ... 12 common frames omitted

```

**Comment by danmason** [comment]
Pushed a couple of commits on here:

- Allow the automatic creation of a single partition event log if specified within the configuration.
- A commit with documentation changes, including general information on how to configure/use the log implementation,

File: modules/azure/README.adoc [hunk 1]
```diff
@@ -59,3 +60,81 @@ These are the following parameters that can be passed within the config for our
...
+| The retention period of the event hub for the Log - the maximum period determined by the tier of eventhub namespace you are using, see the https://learn.microsoft.com/en-us/azure/event-hubs/event-hubs-features#event-retention[Azure Docs].
+| No - only needed if creating event hub automatically, and defaults to "7".
+
+|=== 
+
+### Using Event Hubs 
+
+Some things to note when setting up event hubs for XTDB:
+
+* When using EventHubs as a log - you will require a pre-existing Event Hubs namespace, see
```
Comment (jarohen) [inline modules/azure/README.adoc]: see ...? :slightly_smiling_face:

File: modules/azure/README.adoc [hunk 2]
```diff
@@ -59,3 +60,81 @@ These are the following parameters that can be passed within the config for our
...
+
+* When using EventHubs as a log - you will require a pre-existing Event Hubs namespace, see 
+* As a bare minimum, whichever credentials you use to authenticate to Azure for the app will require two roles on the namespace - *Azure Event Hubs Data sender* and *Azure Event Hubs Data receiver*. 
+
+#### Creating the Event Hub Manually
+
+When creating an eventhub manually to use as an XTDB log, there are a few properties to consider:
+
+* Partition count should be set to *1* - XTDB will only ever use a single partition within it's implementations of Log as they are required to be *totally ordered*.
+* The retention period is configurable - you will likely wish to set this to as high as you reasonably can given the priving tier of your Event Hubs namespace.
```
Comment (jarohen) [inline modules/azure/README.adoc]: pricing?

File: modules/azure/README.adoc [hunk 3]
```diff
@@ -59,3 +60,81 @@ These are the following parameters that can be passed within the config for our
...
+When creating an eventhub manually to use as an XTDB log, there are a few properties to consider:
+
+* Partition count should be set to *1* - XTDB will only ever use a single partition within it's implementations of Log as they are required to be *totally ordered*.
+* The retention period is configurable - you will likely wish to set this to as high as you reasonably can given the priving tier of your Event Hubs namespace. 
+
+#### Creating the Event Hub Automatically
+
+If `create-event-hub?` is set to `true`, XTDB will attempt to create an Event Hub on your behalf - some notes on this:
+
+* In `create-event-hub?` is set, a number of other pieces of configuration must be done to allow your application to manage event hubs on your behalf:
```
Comment (jarohen) [inline modules/azure/README.adoc]: If?

File: modules/azure/README.adoc [hunk 4]
```diff
@@ -59,3 +60,81 @@ These are the following parameters that can be passed within the config for our
...
+** Whichever credentials you use to authenticate to Azure for the app will require one extra roles for the namespace - *Azure Event Hubs Data Owner*.
+** Within your XTDB integrant config, you will need to directly specify the resource group the event hub namespace belongs to within the `resource-group-name` parameter.
+** The application will require two azure related environment variables to be set - `AZURE_SUBCRIPTION_ID` & `AZURE_TENANT_ID`. See the https://learn.microsoft.com/en-us/azure/azure-portal/get-subscription-tenant-id[Azure docs] for more info.
```
Comment (jarohen) [inline modules/azure/README.adoc]: SUBSCRIPTION_ID?

Comment (jarohen) [inline modules/azure/README.adoc]: nit: consistent capitalisation for Azure and Event Hub throughout

File: modules/azure/src/main/clojure/xtdb/azure/log.clj [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      (try
+        (->> (.receiveFromPartition consumer "0" limit event-position max-wait-time)
+             (mapv ->log-record))
+        (catch RuntimeException e
+          (if-let [cause (.getCause e)]
+            (throw cause)
+            (throw e))))))
+
+  (subscribe [this after-tx-id subscriber]
+    (log/handle-polling-subscription this after-tx-id {:poll-sleep-duration (Duration/ofSeconds 1)} subscriber))
```
Comment (jarohen) [inline modules/azure/src/main/clojure/xtdb/azure/log.clj]: can this `PT1S` be configurable?

**Review by jarohen — APPROVED** [review]
nice one @danmason :pray: only nits and typos below, feel free to squash + merge once you've fixed them :slightly_smiling_face:

**Pull request merged**